### PR TITLE
Ensure `Enumerable#to_a` and `Enumerable#tally` properly retain return type of `T`

### LINKED ISSUE #14440: `Enumerable(T)#to_a` regression with interface type
It seems https://github.com/crystal-lang/crystal/pull/12653 introduced a regression when calling `#to_a` on an enumerable of a module type. For example,

```cr
private struct SomeStruct(T, S)
  include Indexable(T)

  @values = [One.new, Two.new]

  def size : Int32
    S
  end

  # # :nodoc:
  def unsafe_fetch(index : Int) : T
    case index
    when 0 then @values[0]
    when 1 then @values[1]
    else
      raise ""
    end
  end
end

module SomeInterface; end

record One do
  include SomeInterface
end

record Two do
  include SomeInterface
end

pp SomeStruct(SomeInterface, 2).new.to_a
```

Works fine on 1.10.0
```sh
$ docker run --rm -it -w /app -v $PWD:/app crystallang/crystal:1.10.0-alpine crystal test.cr
[One(), Two()]
```

Fails on 1.11.0
```sh
$ docker run --rm -it -w /app -v $PWD:/app crystallang/crystal:1.11.0-alpine crystal test.cr
Showing last frame. Use --error-trace for full trace.

In /usr/share/crystal/src/enumerable.cr:2000:14

 2000 | def to_a : Array(T)
                   ^
Error: method SomeStruct(SomeInterface, 2)#to_a must return Array(SomeInterface) but it is returning Array(One | Two)
```

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Fixes #14440

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
Welp, seems wasm doesn't like this new spec...

File: spec/std/indexable_spec.cr [hunk 1]
```diff
@@ -1,6 +1,35 @@
...
+
+  def unsafe_fetch(index : Int) : T
+    case index
+    when 0 then @values[0]
+    when 1 then @values[1]
+    else
+      raise ""
+    end
+  end
+end
```
Comment (straight-shoota) [inline spec/std/indexable_spec.cr:31]: **suggestion:** This helper type could be simplified quite a bit. There's no need for generic arguments or instance variables.
```suggestion
private struct InterfaceIndexable
  include Indexable(SomeInterface)
  
  def size
    2
  end

  def unsafe_fetch(index : Int) : SomeInterface
    case index
    when 0 then One.new
    when 1 then Two.new
    else
      raise ""
    end
  end
end
```

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
Yeah, something's really funky about the failures on macos and wasm.

**Comment by Blacksmoke16** [comment]
I'm actually able to reproduce this on my intel mac with LLVM 17. Reproduces when running the whole of `std_spec`, or both specs together, but not when running the new modified specs on their own.

I'll play around with it and see if I can get a more minimal example...

**Comment by Blacksmoke16** [comment]
I think the gist of this is it's caused by having two `private` types typed as an interface interfering with each other when required from separate files. 

E.g. 
```cr
# test2.cr
module OtherInterface; end

private record One do
  include OtherInterface
end

private record Two do
  include OtherInterface
end

([One.new, Two.new] of OtherInterface) == [One.new, Two.new]
```
Then:
```cr
# test.cr
require "./test2"

module OtherInterface; end

private record One do
  include OtherInterface
end

private record Two do
  include OtherInterface
end

pp [One.new, Two.new] == [One.new, Two.new]
```
Results in:
```sh
$ crystal test.cr
Program exited because of an invalid instruction
```
If I had to guess, I'd say something regarding the "private to the file" thing isn't working out exactly as we expect giving one or the other files a reference to the other?

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**